### PR TITLE
Make GlobalCtxt implement Sync

### PR DESCRIPTION
**Pull request by Zoxc**
This PR makes the `GlobalCtxt` implement `Sync`, which is the first step towards making rustc thread safe so we can execute queries in multiple threads.

The changes here are mostly mechanical where `Cell` is replaced by `LockCell`, `Rc` is replaced by `Lrc` and `RefCell` is replaced by either `Lock` or `RwLock`. Some thread local variables are now declared with `rustc_global!`, which indicates that they are intended to be global to a rustc session, however they aren't changed to global variables since rustdoc has multiple threads with rustc sessions in them.

Here is a list of the larger changes:
- `src/librustc_data_structures/sync.rs` has the new types other items related to thread safety.
- `src/libarena/lib.rs` is changed so that there's a lock around each arena if thread safety is enabled.
- There's a new `[rust] experimental_parallel_queries` entry for `config.toml` which enables thread safety. It is disabled by default.
- `-Z query_threads` allows you to set the number of thread used to compute queries, although it is current ignored.
- It uses my fork of [owning_ref](https://github.com/Zoxc/owning-ref-rs), so we can erase a `OwningRef` into a thread-safe type.
- `GlobalArenas` and `DroplessArena` are combined into `AllArenas` which is the type drivers shoud use. This refactoring makes it easier to create thread-local arenas in the future.
- A `SameThread` type is introduces in `src/librustc_trans/metadata.rs` which asserts that operations on the inner type all happen on the same thread. This is used to make `ArchiveRO` and `ObjectFile` `Send + Sync` so we can make the `[u8]` metadata reference into them `Send + Sync`.
- `ThreadCtxt` is introduced. `TyCtxt` now points to it instead of `GlobalCtxt`. `ThreadCtxt` contains a references to a `GlobalCtxt` and contains the parts of the `GlobalCtxt which has to be thread local.
- `mk_attr_id` in `src/libsyntax/attr.rs` is changed to use an `AtomicUsize`.
- The `err_count` field of `Handler` in `src/librustc_errors/lib.rs` is now a `AtomicUsize` instead of `Cell<usize>`.

One thing to note is that `Lock` and `RwLock` implements `Clone`, since there's code which relies the `Clone` impl for `RefCell`. There's some risk of deadlocks due to that. Replacing `RefCell` with `Lock` also risks deadlocks, since `RefCell` is re-entrant for read-locks, while `Lock` is not.

### CODE REVIEW

**Comment by shepmaster** [comment]
Assigning to...

r? @pnkfelix

**Comment by Zoxc** [comment]
What is everyone opinion on introducing locks and `Arc`s in rustc in an effort to make GlobalCtxt thread safe? We could redirect these types to `RefCell`, `Rc`, etc. in production builds to avoid paying the overhead for these thread safe types.

cc @rust-lang/compiler

**Comment by eddyb** [comment]
I would rather not introduce `#[cfg]`s into `rustc` like that, FWIW.

**Comment by Zoxc** [comment]
I'd like the ability to turn on and off thread safety so we can easily measure the overhead of it. That is going to require some `#[cfg]`s at least.

**Comment by michaelwoerister** [comment]
If the `#[cfg]`s are very localized and they are clearly experimental, I'm not strictly against adding some. However, in such a case the effort of keeping such a patch around locally would not be too big either.

**Comment by Zoxc** [comment]
@Mark-Simulacrum Can I get a perf run when bors completes?

**Comment by kennytm** [comment]
@Zoxc please break every tool in `src/tools/toolstate.toml` and try again.

File: src/libarena/Cargo.toml [hunk 1]
```diff
@@ -7,3 +7,6 @@ version = "0.0.0"
 name = "arena"
 path = "lib.rs"
 crate-type = ["dylib"]
+
+[dependencies]
+rustc_data_structures = { path = "../librustc_data_structures" }
```
Comment (bjorn3) [inline src/libarena/Cargo.toml]: Missing newline

File: src/librustc_data_structures/Cargo.toml [hunk 2]
```diff
@@ -11,3 +11,8 @@ crate-type = ["dylib"]
 [dependencies]
 log = "0.3"
 serialize = { path = "../libserialize" }
+owning_ref = "0.3"
+
+[dependencies.parking_lot]
+version = "0.5"
+features = ["nightly", "owning_ref"]
```
Comment (bjorn3) [inline src/librustc_data_structures/Cargo.toml]: Missing newline

File: src/librustc_data_structures/lock.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        self.lock()
+    }
+}
+
+impl<T: Clone> Clone for Lock<T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        Lock::new(self.borrow().clone())
+    }
+}
```
Comment (bjorn3) [inline src/librustc_data_structures/lock.rs]: Missing newline

**Review by bjorn3 — CHANGES_REQUESTED** [review]

**Comment by Zoxc** [comment]
Currently there's some cfg's related to adding `Send` and `Sync` on trait objects. I expect these changes to go away when the next beta is used to build rustc. Please ignore them for now. Also see the initial comments for a overview of the changes.

r? @arielb1

**Comment by Zoxc** [comment]
@Mark-Simulacrum Can I get a perf run on the try build which just completed?

**Comment by Mark-Simulacrum** [comment]
I've started perf on the try commit, it should be live in ~30 minutes or so. The comparison page should just work.

**Comment by kennytm** [comment]
The comparison page FYI. http://perf.rust-lang.org/compare.html?start=e06138338f78cd2695f857855e685e9e50a2486b&end=90ded1e1ad7c914726b74b2fa8afbd0794f47709&stat=instructions%3Au. Basically red in every metric.

**Comment by arielb1** [comment]
@kennytm 

This PR is just adding locks, so of course it would regress performance.

**Comment by Zoxc** [comment]
The result for `regex-opt.0.1.30` is strange. It should spend more time in LLVM relative to `regex.0.1.30` so the overhead of the locking should decrease. Yet it has almost 3x the overhead.

**Comment by Zoxc** [comment]
@Mark-Simulacrum Could you also run perf on the `699cd27b9c19fa921732aa45644e3f1295066ab9` try build?

**Comment by Mark-Simulacrum** [comment]
Started; same as previously expect ~30 minute delay.

**Comment by Zoxc** [comment]
[Here are the results](http://perf.rust-lang.org/compare.html?start=c5c70ef72369b00870d9fc67b99853b0681e17e2&end=699cd27b9c19fa921732aa45644e3f1295066ab9&stat=instructions%3Au). This build has all the thread safety disabled, so all the tests should be close to 0%, yet only a few are.

**Comment by shepmaster** [comment]
Ping from triage @Zoxc — we haven't heard from you in a few days! Will you be able to address your conflicts?

**Comment by Zoxc** [comment]
Since the new bootstrap compiler landed, all the cfg's are now inside the `sync` module :)

I don't know what to do with the `owning_ref` fork though. Do we have a policy of what to do with upstream crate forks?

**Comment by Zoxc** [comment]
This is being split into smaller PRs.

**Pull request closed without merging**